### PR TITLE
TRAD-183 Go SDK update with recent block hash api

### PR DESCRIPTION
**Pull request by yobloxroute**

### CODE REVIEW

File: provider/ws.go [hunk 1]
```diff
@@ -12,9 +12,10 @@ import (
 type WSClient struct {
 	pb.UnimplementedApiServer
 
-	addr       string
-	conn       *connections2.WS
-	privateKey *solana.PrivateKey
+	addr                 string
+	conn                 *connections2.WS
```
Comment (aspin) [inline provider/ws.go]: can you do a bit of refactoring and delete this alias?

File: provider/recent_hash_store.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		select {
+		case hash := <-ch:
+			s.updateBlockHash(hash)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *recentBlockHashStore) updateBlockHash(hash *pb.GetRecentBlockHashResponse) {
```
Comment (aspin) [inline provider/recent_hash_store.go]: can simplify all the functions in this struct's name -- it's already called `recentBlockHashStore`, so the `BlockHash` is implied. so methods can be `update()`, `get()` and `cached()`

File: provider/recent_hash_store.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		hashProvider:       hashProvider,
+		hashStreamProvider: streamProvider,
+		hash:               "",
+		hashTime:           time.Time{},
+		hashExpiryDuration: opts.CacheBlockHash,
+	}
+}
+
+func (s *recentBlockHashStore) run() {
+	ctx := context.Background()
```
Comment (aspin) [inline provider/recent_hash_store.go]: should probably pass this context through the store when initializing it, as there's currently no way to cancel this loop

File: provider/recent_hash_store.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	response := s.getCachedBlockHash()
+	if response != nil {
+		return response, nil
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	now := time.Now()
+	if s.hash != "" && s.hashTime.Before(now.Add(s.hashExpiryDuration)) {
```
Comment (aspin) [inline provider/recent_hash_store.go]: condition isn't necessary: `getCachedBlockHash` already checks this conditions

File: provider/recent_hash_store.go (hunk 3 above)
Comment (aspin) [reply provider/recent_hash_store.go]: I'm ok with requiring a `ctx` object for each provider type

File: provider/grpc.go [hunk 5]
```diff
@@ -385,3 +402,13 @@ func (g *GRPCClient) GetOrderStatusStream(ctx context.Context, market, ownerAddr
...
 }
+
+// GetRecentBlockHashStream subscribes to a stream for getting recent block hash.
+func (g *GRPCClient) GetRecentBlockHashStream(ctx context.Context) (connections2.Streamer[*pb.GetRecentBlockHashResponse], error) {
+	stream, err := g.apiClient.GetRecentBlockHashStream(ctx, &pb.GetRecentBlockHashRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	return connections2.GRPCStream[pb.GetRecentBlockHashResponse](stream, ""), nil
```
Comment (aspin) [inline provider/grpc.go]: can you also clean up the aliases in this file?

File: provider/common.go [hunk 6]
```diff
@@ -37,10 +37,11 @@ type PostOrderOpts struct {
...
 type RPCOpts struct {
-	Endpoint   string
-	Timeout    time.Duration
-	PrivateKey *solana.PrivateKey
-	AuthHeader string
+	Endpoint       string
+	Timeout        time.Duration
+	PrivateKey     *solana.PrivateKey
+	AuthHeader     string
+	CacheBlockHash time.Duration
```
Comment (aspin) [inline provider/common.go]: should also have a parameter to enable this. by default, the `recentBlockHashStore` should not run

**Review by aspin — CHANGES_REQUESTED** [review]

**Review by Uri-Margalit-bloXroute — COMMENTED** [review]
as a first timer seen those call methods it seems strange that the function returns `true` if it fails and `false` if succeeded .

The `failed = failed || function()` can be changed to `failed = failed && function()` and the functions can return `true` if succeeded

File: provider/ws.go (hunk 1 above)
Comment (yobloxroute) [reply provider/ws.go]: you mean connections2? yeah, it looks strange will do

Comment (yobloxroute) [reply provider/ws.go]: done

File: provider/recent_hash_store.go (hunk 2 above)
Comment (yobloxroute) [reply provider/recent_hash_store.go]: done

File: provider/recent_hash_store.go (hunk 3 above)
Comment (yobloxroute) [reply provider/recent_hash_store.go]: yeah, in general it's worth even to pass context when creating client, that's not a part of the api

File: provider/recent_hash_store.go (hunk 4 above)
Comment (yobloxroute) [reply provider/recent_hash_store.go]: done

File: provider/grpc.go (hunk 5 above)
Comment (yobloxroute) [reply provider/grpc.go]: done

File: provider/common.go (hunk 6 above)
Comment (yobloxroute) [reply provider/common.go]: agree, changed

**Comment by yobloxroute** [comment]
> as a first timer seen those call methods it seems strange that the function returns `true` if it fails and `false` if succeeded .
> 
> The `failed = failed || function()` can be changed to `failed = failed && function()` and the functions can return `true` if succeeded

believe it was done to accumulate error, but yeah, looks strange

**Review by aspin — APPROVED** [review]

**Pull request merged**